src/drivers/amd/common/acp_dmic_dma.c

```diff
@@ -106,10 +107,39 @@ static int acp_dmic_dma_status(struct dma_chan_data *channel,
 static int acp_dmic_dma_copy(struct dma_chan_data *channel, int bytes,
 			     uint32_t flags)
 {
+	uint32_t i;
+	uint32_t j;
+	int numsamples;
+	char *dmic_rngbuff_addr2 = initsilence.dmic_rngbuff_addr1;
 	struct dma_cb_data next = {
 		.channel = channel,
 		.elem.size = bytes,
 	};
+	if (initsilence.silence_incr < initsilence.silence_cnt) {
+		if (initsilence.silence_incr % 2)
+			dmic_rngbuff_addr2 = initsilence.dmic_rngbuff_addr1 +
+								(dmic_rngbuff_size >> 1);
+		for (i = 0; i < (dmic_rngbuff_size >> 1); i = i + 1)
+			dmic_rngbuff_addr2[i] = 0;
+		initsilence.silence_incr = initsilence.silence_incr + 1;
+	} else if (initsilence.silence_incr < (initsilence.silence_cnt +
+			initsilence.numfilterbuffers)) {
+		numsamples = ((dmic_rngbuff_size >> 1) / (initsilence.num_chs *
+					initsilence.bytes_per_sample));
+		if (initsilence.silence_incr % 2)
+			dmic_rngbuff_addr2 = initsilence.dmic_rngbuff_addr1 +
+								(dmic_rngbuff_size >> 1);
+		initsilence.dmic_rngbuff_iaddr = (int *)dmic_rngbuff_addr2;
+		for (i = 0; i < (numsamples * initsilence.num_chs);
+			 i = i + initsilence.num_chs, initsilence.coeff++) {
+			for (j = 0; j < initsilence.num_chs; j++) {
```

Comment:
I think these nested arrays can be a single one something like
```
for (i = 0; i < numsamples * initsilence.num_chs; i++)
    initsilence.dmic_rngbuff_iaddr[i] = ...
```

Reply:
nested for loop seems to be proper.

Reply:
sorry, could you explain "proper?" Is my proposal wrong or you find your code better for some other reason?

Reply:
ah, ok, you also increment `initsilence.coeff` in the external loop, then yes, merging them into a single loop wouldn't be that easy.